dlt/destinations/impl/sink/sink.py

```diff
@@ -0,0 +1,170 @@
+            assert not table_update
+            coerced_items.append(coerced_item)
+
+        # send single item on batch size 1
+        if self._config.batch_size == 1:
+            coerced_items = coerced_items[0]
+
+        # call callable
+        self._config.credentials.resolved_callable(coerced_items, self._table)
+
+    def state(self) -> TLoadJobState:
+        return self._state
+
+    def exception(self) -> str:
+        raise NotImplementedError()
+
+
+class SinkParquetLoadJob(SinkLoadJob):
+    def run(self, start_index: int) -> Iterable[TDataItems]:
+        # stream items
+        from dlt.common.libs.pyarrow import pyarrow
+
+        # guard against changed batch size after restart of loadjob
+        assert (
+            start_index % self._config.batch_size
+        ) == 0, "Batch size was changed during processing of one load package"
+
+        start_batch = start_index / self._config.batch_size
+        with pyarrow.parquet.ParquetFile(self._file_path) as reader:
+            for record_batch in reader.iter_batches(batch_size=self._config.batch_size):
+                if start_batch > 0:
+                    start_batch -= 1
+                    continue
+                batch = record_batch.to_pylist()
+                yield batch
+
+
+class SinkJsonlLoadJob(SinkLoadJob):
+    def run(self, start_index: int) -> Iterable[TDataItems]:
+        current_batch: TDataItems = []
+
+        # stream items
+        with FileStorage.open_zipsafe_ro(self._file_path) as f:
+            for line in f:
+                # find correct start position
+                if start_index > 0:
+                    start_index -= 1
+                    continue
+                current_batch.append(json.loads(line))
+                if len(current_batch) == self._config.batch_size:
+                    yield current_batch
+                    current_batch = []
+            yield current_batch
+
+
+class SinkClient(JobClientBase):
+    """Sink Client"""
+
+    capabilities: ClassVar[DestinationCapabilitiesContext] = capabilities()
+
+    def __init__(self, schema: Schema, config: SinkClientConfiguration) -> None:
+        super().__init__(schema, config)
+        self.config: SinkClientConfiguration = config
+
+    def initialize_storage(self, truncate_tables: Iterable[str] = None) -> None:
+        pass
+
+    def is_storage_initialized(self) -> bool:
+        return True
+
+    def drop_storage(self) -> None:
+        pass
+
+    def update_stored_schema(
+        self, only_tables: Iterable[str] = None, expected_update: TSchemaTables = None
+    ) -> Optional[TSchemaTables]:
+        return super().update_stored_schema(only_tables, expected_update)
+
+    def start_file_load(self, table: TTableSchema, file_path: str, load_id: str) -> LoadJob:
+        # save our state in destination name scope
+        load_state = load_package_state().setdefault(self.config.destination_name, {})
```

Comment:
now when I think of it we do not need injection at all. you have load id so create PackageStorage here and pass it to sink load job. then use the method directly. (don't forget to lock on commit)

Reply:
hm, are you sure? if we get the current loadpackagestate from the state file in every thread from the storage, we have multiple copies in memory and it will not be thread safe anymore even if we lock on commit. we'd have to lock, then load update dict and commit and then release the lock. Or am i misunderstanding what you are suggesting?

Reply:
Alright, if I just keep a copy of the state in memory it should not be a problem.